src/components/shapes/draw.js

```diff
@@ -579,6 +599,224 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
+    var xanchor = textPos.xanchor;
+
+    function textLayout(s) {
+        s.call(Drawing.font, font)
+            .attr({
+                'text-anchor': {
+                    left: 'start',
+                    center: 'middle',
+                    right: 'end'
+                }[xanchor],
+                'y': texty,
+                'x': textx,
+                'transform': 'rotate(' + textangle + ',' + textx + ',' + texty + ')'
+            });
+        svgTextUtils.convertToTspans(s, gd);
+        return s;
+    }
+
+    labelText.call(textLayout);
+}
+
+function calcTextAngle(shapex0, shapey0, shapex1, shapey1) {
+    var dy, dx;
+    dx = Math.abs(shapex1 - shapex0);
+    if(shapex1 >= shapex0) {
+        dy = shapey0 - shapey1;
+    } else {
+        dy = shapey1 - shapey0;
+    }
+    return -180 / Math.PI * Math.atan2(dy, dx);
+}
+
+function calcTextPosition(shapex0, shapey0, shapex1, shapey1, shapeOptions, actualTextAngle, textBB) {
+    var textPosition = shapeOptions.label.textposition;
+    var textPadding = shapeOptions.label.padding;
+    var shapeType = shapeOptions.type;
+    var textAngleRad = Math.PI / 180 * actualTextAngle;
+    var xanchor = shapeOptions.label.xanchor;
+    var yanchor = shapeOptions.label.yanchor;
+
+    var textx, texty, paddingX, paddingY;
+
+    // Text position functions differently for lines vs. other shapes
+    if(shapeType === 'line') {
+        // Set base position for start vs. center vs. end of line (default is 'center')
+        if(textPosition.indexOf('start') !== -1) {
+            textx = shapex0;
+            texty = shapey0;
+            if(xanchor === 'auto') xanchor = (shapex1 >= shapex0) ? 'right' : 'left';
+        } else if(textPosition.indexOf('end') !== -1) {
+            textx = shapex1;
+            texty = shapey1;
+            if(xanchor === 'auto') xanchor = (shapex1 >= shapex0) ? 'left' : 'right';
+        } else { // Default: center
+            textx = (shapex0 + shapex1) / 2;
+            texty = (shapey0 + shapey1) / 2;
+            if(xanchor === 'auto') xanchor = 'center';
+        }
+
+        // Default yanchor is 'bottom' for lines (should handle in defaults.js)
+        if(yanchor === 'auto') yanchor = 'bottom';
+
+        // Special case for padding when angle is 'auto' for lines
+        // Padding should be treated as an orthogonal offset in this case
+        // Otherwise, padding is just a simple x and y offset
+        if(shapeOptions.label.textangle === 'auto') {
+            // Set direction to apply padding (based on `yanchor` only)
+            var paddingDirection = { bottom: 1, middle: 0, top: -1 }[yanchor];
+            paddingX = textPadding * Math.sin(textAngleRad) * paddingDirection;
+            paddingY = -textPadding * Math.cos(textAngleRad) * paddingDirection;
+        } else {
+            // Set direction to apply padding (based on `xanchor` and `yanchor`)
+            var paddingDirectionX = { left: 1, center: 0, right: -1 }[xanchor];
```

Comment:
Once you addressed https://github.com/plotly/plotly.js/pull/6454/files#r1121818176,
I may suggest to declare these constant objects once in the upper scope, reuse or create an object similar to https://github.com/plotly/plotly.js/blob/master/src/constants/alignment.js